File: src/variant_1/UserSession.java
package variant_1;

import common.User;
import common.UserRepository;

/*
* Класс, реализующий имитацию деятельности пользователя.
* */
public class UserSession implements Runnable {
    private MyThreadLocal<User> thlUser = new MyThreadLocal();
    private MyThreadLocal<String> thlThreadInfo = new MyThreadLocal();
    private MyThreadLocal<Long> thlCounter = new MyThreadLocal();

    // Id пользователя, который "открывает сессию".
    private int userId;
    private UserRepository repo = new UserRepository();
    private long operations;

    public UserSession(int userId, long operations) {
        this.userId = userId;
        this.operations = operations;
    }

    @Override
    public void run() {
        User user = repo.getUser(userId);

        thlUser.put(user);
        thlThreadInfo.put("The pid is " + Thread.currentThread().getId() + "; thread name is " + Thread.currentThread().getName());
        thlCounter.put((long) 1);

        while (thlCounter.get() < operations) {
            thlCounter.put(thlCounter.get() + 1);
        }

        System.out.println(userId + " — User: " + thlUser.get().toString());
        System.out.println(userId + " — Thread info: " + thlThreadInfo.get());
        System.out.println(userId + " — Number of ops: " + thlCounter.get());
    }
}

File: src/context/UserSession.kt
package context

import storage.MyThreadLocal

class UserSession : Runnable {
    private var userId: Int // Идентификатор пользователя.
    private var localStorage: ThreadLocal<User> // LocalStorage из стандартной java-библиотеки.
    private var myLocalStorage: MyThreadLocal<User> // Собственная реализация класса ThreadLocal.

    constructor(id: Int) {
        userId = id
        localStorage = ThreadLocal()
        myLocalStorage = MyThreadLocal()
    }

    override fun run() {
        var repository = UserRepository()
        var userData = repository.getUser(userId) as User

        localStorage.set(userData)
        localStorage.set(User(0, "test", 10, arrayListOf("1", "2", "3")))
        localStorage.get()
        localStorage.remove()

        /*userData.age = (Thread.currentThread().id).toInt()
        myLocalStorage.set(userData)
        println("Started session for $userId")
        println("Session closed! ${myLocalStorage.get()?.name} of age ${myLocalStorage.get()?.age}, good bye!")*/
    }
}